データベースのテーブル設計

tonakai-it opened the pull request:
# what
テーブルの設計
# why
ユーザー、アイテム、purchase(購入)、アドレス(発送先)の４つのテーブル

### 修正
ユーザーとアドレスのカラム追加
購入とアドレスのアソシエーション正常化
reference型のカラム名の後の_id削除
アクティブハッシュのカラム名の後に_id追加
電話番号のカラム型をstringに

以下、ER図
https://gyazo.com/bc33b978c96eccba7f34ae933386f61e
↓
【修正後】
https://gyazo.com/583acf833c0a1d513eaadf065800b9d0

tonakai-it committed:
データベースのテーブル設計

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,58 @@
+| email     | string  | null: false               |
+| password  | string  | null: false               |
+| birthday  | integer | null: false               |
+| real_name | string  | null: false               |
```
| real_name | string  | null: false               |

ユーザーの登録時、姓、名、姓（フリガナ）、名（フリガナ）と入力欄が細かく分かれているはずです。
カラムは姓、名、姓（フリガナ）、名（フリガナ）とそれぞれ設計しましょう！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,58 @@
+| nickname  | string  | null: false, unique: true |
+| email     | string  | null: false               |
+| password  | string  | null: false               |
+| birthday  | integer | null: false               |
```
| birthday  | integer | null: false               |

こちらのTypeは「date型」を用いることで年、月、日を1つにして保存できます。

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,58 @@
+| --------- | ------- | ------------------------- |
+| nickname  | string  | null: false, unique: true |
+| email     | string  | null: false               |
+| password  | string  | null: false               |
```
| password | string  | null: false |

こちらのカラムはdeviseのgemを使用しての実装になるため、deviseが予め用意している
encrypted_passwordというカラムを用意してあげましょう。

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,58 @@
+| Column    | Type    | Options                   |
+| --------- | ------- | ------------------------- |
+| nickname  | string  | null: false, unique: true |
+| email     | string  | null: false               |
```
| email | string | null: false | //メールアドレス
下記のユーザー管理機能の実装条件に合うように、Optionsに記述を追記しましょう。

- メールアドレスが一意性であること

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,58 @@
+| ------------ | --------- | ------------------------------------------- |
+| name         | string    | null: false                                 |
+| introduction | text      | null: false                                 |
+| item_profile | text      | null: false                                 |
```
```
| introduction | text      | null: false                                 |
| item_profile | text      | null: false                                 |
```
これらどちらも商品説明のカラムでしょうか？その様でありましたら、カラムを１つにいたしましょう！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,58 @@
+| name         | string    | null: false                                 |
+| introduction | text      | null: false                                 |
+| item_profile | text      | null: false                                 |
+| category     | string    | null: false                                 |
+| status       | string    | null: false                                 |
```
```
| category     | string    | null: false                                 |
| status       | string    | null: false                                 |
```
これら意外にも** 配送料の負担, 発送元の地域,  発送までの日数**のカラムも必要かと思いますので追加いたしましょう。
また、**商品の状態, 配送料の負担, 発送元の地域,  発送までの日数, カテゴリー**のカラムはactive_hashで実装の予定のため、integer型にしましょう！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,58 @@
+## purchases テーブル
+| Column     | Type      | Options                         |
+| ---------- | --------- | ------------------------------- |
+| address_id | reference | null: false, foreign_key: true  |
```
| address_id | reference | null: false, foreign_key: true  |

購入情報は１つの住所を持っている設計にしたいので、purchases テーブルには住所の外部キーは不要となります。
また、住所とのアソシエーションはbelongs_toではなくhas_oneといたしましょう！

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,58 @@
+| user_id | reference | null: false, foreign_key: true  |
+| item_id | reference | null: false, foreign_key: true  |
+
+- belongs_to :user
+- belongs_to :item
```
今回の設計では「配送先住所と購入管理が紐づく」実装を行なっていきたいので、商品やユーザーと配送先住所のアソシエーションは不要となります。

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,58 @@
+
+- belongs_to :user
+- belongs_to :item
+- has_one :purchase
```
>has_one :purchase

こちらbelongs_to :purchaseになります！
-「配送先住所は1つの購入履歴に属している」という設計を行いたいため。

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,58 @@
+| Column  | Type      | Options                         |
+| ------- | --------- | ------------------------------- |
+| address | text      | null:false                      |
+| user_id | reference | null: false, foreign_key: true  |
+| item_id | reference | null: false, foreign_key: true  |
```
```
| user_id | reference | null: false, foreign_key: true  |
| item_id | reference | null: false, foreign_key: true  |
```

ユーザーや商品の外部キーカラムではなくpurchasesの外部キーカラムを用意いたしましょう。

Katsuno0814 commented on README.md:
```diff
@@ -1,24 +1,58 @@
+## addresses テーブル
+| Column  | Type      | Options                         |
+| ------- | --------- | ------------------------------- |
+| address | text      | null:false                      |
```
>| address | text      | null:false                      |

こちら配送先住所を入力するさいに必要な市町村、番地、建物番号、郵便番号、電話番号のカラムが必要になります。
見本アプリを参考に何の情報が必要か見直してみましょう！

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

tonakai-it committed:
訂正１

add column "shipping_fee etc..",and"village_number etc..." to users or address table

correct assosiation between purchase and address

delete assosiation address table with user or item

any other adjastment

Katsuno0814 commented on README.md:
```diff
@@ -4,55 +4,61 @@
+| send_from     | integer   | null: false                                 |
+| shipping_days | integer   | null: false                                 |
+| price         | integer   | null: false                                 |
+| user_id       | reference | reference | null: false, foreign_key: true  |
```
>| user_id       | reference | reference | null: false, foreign_key: true  |

references 型で記述する際は、カラム名の_idは不要となります。
他テーブルでも同じ様な記述がありましたら修正お願いいたします。

Katsuno0814 commented on README.md:
```diff
@@ -4,55 +4,61 @@
-| item_id | reference | null: false, foreign_key: true  |
+| Column          | Type      | Options                         |
+| --------------- | --------- | ------------------------------- |
+| post_number     | integer   | null:false                      |
```
>| post_number     | integer   | null:false                      |

こちら| post_number     | string   | null:false                      |ですね！

Katsuno0814 commented on README.md:
```diff
@@ -4,55 +4,61 @@
+| village_number  | string    | null:false                      |
+| building_detail | string    | null:false                      |
+| tele_number     | integer   | null:false                      |
+| user_id         | reference | null: false, foreign_key: true  |
+| item_id         | reference | null: false, foreign_key: true  |
```
>| user_id         | reference | null: false, foreign_key: true  |
| item_id         | reference | null: false, foreign_key: true  |

こちらpurchases テーブルを作成されておりますので不要となります、削除いたしましょう。

Katsuno0814 commented on README.md:
```diff
@@ -4,55 +4,61 @@
+| Column          | Type      | Options                         |
+| --------------- | --------- | ------------------------------- |
+| post_number     | integer   | null:false                      |
+| prefecture      | integer   | null:false                      |
```
>| prefecture      | integer   | null:false                      |

こちら| prefecture_id      | integer   | null:false                      |となります！

Katsuno0814 commented on README.md:
```diff
@@ -4,55 +4,61 @@
+| prefecture      | integer   | null:false                      |
+| village_name    | string    | null:false                      |
+| village_number  | string    | null:false                      |
+| building_detail | string    | null:false                      |
```
| building_detail | string    | null:false                      |

今回の実装では、建物番号の入力は任意になるため、not null制約は不要となります。

Katsuno0814 commented on README.md:
```diff
@@ -4,55 +4,61 @@
+| village_name    | string    | null:false                      |
+| village_number  | string    | null:false                      |
+| building_detail | string    | null:false                      |
+| tele_number     | integer   | null:false                      |
```
| tele_number     | integer   | null:false                      |

電話番号は数字の文字列として実装しておきたいためstring型で実装しましょう。

Katsuno0814 commented on README.md:
```diff
@@ -4,55 +4,61 @@
+| ------------- | --------- | ------------------------------------------- |
+| name          | string    | null: false                                 |
+| introduction  | text      | null: false                                 |
+| category      | integer   | null: false                                 |
+| status        | integer   | null: false                                 |
+| shipping_fee  | integer   | null: false                                 |
+| send_from     | integer   | null: false                                 |
+| shipping_days | integer   | null: false                                 |
```
```
| category      | integer   | null: false                                 |
| status        | integer   | null: false                                 |
| shipping_fee  | integer   | null: false                                 |
| send_from     | integer   | null: false                                 |
| shipping_days | integer   | null: false                                 |

```
activehashのカラム名は「単数形+_id」になりますので修正お願いいたします！

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

tonakai-it committed:
修正2

reference型のカラム名の後の_id削除
アクティブハッシュのカラム名の後に_id追加
電話番号のカラム型をstringに

tatsuki-kato commented on README.md:
```diff
@@ -1,24 +1,62 @@
+| shipping_fee_id  | integer   | null: false                                 |
+| send_from_id     | integer   | null: false                                 |
+| shipping_days_id | integer   | null: false                                 |
+| price            | integer   | null: false                                 |
+| user             | reference | reference | null: false, foreign_key: true  |
+
```
```md
| user             | reference | reference | null: false, foreign_key: true  |
```
こちら`reference`ではなく、`references`にしましょう！
また、`reference`が２回記述されているため、どちらかを削除しましょう。

tatsuki-kato commented on README.md:
```diff
@@ -1,24 +1,62 @@
+
+## purchases テーブル
+| Column | Type      | Options                         |
+| ---------- | --------- | ------------------------------- |
+| user   | reference | null: false, foreign_key: true  |
+| item   | reference | null: false, foreign_key: true  |
+
```
こちらも同様に`references`にしましょう。

tatsuki-kato commented on README.md:
```diff
@@ -1,24 +1,62 @@
+| village_name    | string    | null:false                      |
+| village_number  | string    | null:false                      |
+| building_detail | string    |                                 |
+| tele_number     | string    | null:false                      |
+| purchase        | reference | null: false, foreign_key: true  |
+
```
```
| purchase        | reference | null: false, foreign_key: true  |
```
こちらも同様に`references`にしましょう。

tatsuki-kato reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、確認をお願いします！

tonakai-it committed:
修正３

表現の修正

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます。

修正点確認いたしました。
LGTMとさせていただきます。実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう。
